File: smartadapter/src/main/java/smartadapter/internal/factory/SmartRecyclerAdapterExtensionFactory.java
package smartadapter.internal.factory;

/*
 * Created by Manne Öhlund on 2019-08-15.
 * Copyright (c) All rights reserved.
 */

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.List;

import smartadapter.SmartExtensionBuilder;
import smartadapter.SmartRecyclerAdapter;
import smartadapter.widget.BasicDragAndDropExtension;
import smartadapter.widget.BasicSwipeExtension;

/**
 * Builds {@link SmartExtensionBuilder} extensions.
 *
 * @see BasicSwipeExtension
 * @see BasicDragAndDropExtension
 */
public class SmartRecyclerAdapterExtensionFactory {

    private List<SmartExtensionBuilder> smartExtensionBuilders = new ArrayList<>();

    public void add(@NonNull SmartExtensionBuilder smartExtensionBuilder) {
        smartExtensionBuilders.add(smartExtensionBuilder);
    }

    public void build(@NonNull SmartRecyclerAdapter smartRecyclerAdapter, @NonNull RecyclerView recyclerView) {
        for (SmartExtensionBuilder smartExtension : smartExtensionBuilders) {
            smartExtension
                    .setSmartRecyclerAdapter(smartRecyclerAdapter)
                    .setRecyclerView(recyclerView)
                    .build();
        }
    }
}

File: smartadapter/src/main/java/smartadapter/internal/factory/SmartRecyclerAdapterExtensionFactory.kt
package smartadapter.internal.factory

/*
 * Created by Manne Öhlund on 2019-08-15.
 * Copyright (c) All rights reserved.
 */

import androidx.recyclerview.widget.RecyclerView
import smartadapter.SmartExtensionBuilder
import smartadapter.SmartRecyclerAdapter
import smartadapter.widget.BasicDragAndDropExtension
import smartadapter.widget.BasicSwipeExtension
import java.util.*

/**
 * Builds [SmartExtensionBuilder] extensions.
 *
 * @see BasicSwipeExtension
 *
 * @see BasicDragAndDropExtension
 */
class SmartRecyclerAdapterExtensionFactory {

    private val smartExtensionBuilders = ArrayList<SmartExtensionBuilder<*>>()

    fun add(smartExtensionBuilder: SmartExtensionBuilder<*>) {
        smartExtensionBuilders.add(smartExtensionBuilder)
    }

    fun build(smartRecyclerAdapter: SmartRecyclerAdapter, recyclerView: RecyclerView) {
        for (smartExtension in smartExtensionBuilders) {
            smartExtension.apply {
                this.smartRecyclerAdapter = smartRecyclerAdapter
                this.recyclerView = recyclerView
            }.build()
        }
    }
}
